### PR TITLE
Interrupt backup

### LINKED ISSUE #2960: Save an "interrupted" snapshot when backup is interrupted by SIGINT
Output of `restic version`
--------------------------
restic 0.10.0 (v0.10.0-23-g506fcfd4-dirty) compiled with go1.14.7 on linux/amd64

What should restic do differently? Which functionality do you think we should add?
----------------------------------------------------------------------------------

When `restic backup` is interrupted by SIGINT, it does not save any snapshot. The only thing that can be used in following backups is the blobs that have been saved and included in an index file that also has been already saved.

This has many drawbacks:

- Users may wonder why snapshots are missing (e.g. for automated backup runs)
- The data already backuped is not accessible (can be only accessed after a `recover` run)
- pack files may have been written which didn't make it in an written index and are effectively "lost"
- a `prune` run will completely remove the interrupted backup which may not be what users wish.

IMO, we can do better! I propose to do the following after getting an SIGINT:

- immediately stop saving of data blobs
- BUT continue saving trees (which should now finish rather quickly)
- finish the index and save it
- also save the snapshot, but with added tag "interrupted"
- if everything of the above does not finish within a given time - e.g. 15s - cancel everything to let restic terminate.

Of course, this approach also involves ignoring snapshots with the tag "interrupted" when search for the parent snapshot.

While I already have a draft implementation which does exactly the steps I pointed out, I would like to use this issue to discuss the solution:

- Is the general way to follow ok?
- Are there some pitfalls I didn't see?
- Should we make the tag name and/or time to hard finish configurable for users?

What are you trying to do? What problem would this solve?
---------------------------------------------------------

See the problems above.

A user in the forum tried to find a workaround and didn't really suceed, see:
https://forum.restic.net/t/forget-instead-of-rebuild-index-then-recover-then-prune/3088

Did restic help you today? Did it make you happy in any way?
------------------------------------------------------------

Not today so far, but makes me happy on several other days :wink:

### PR DESCRIPTION
**Pull request by aawsome**
What does this PR change? What problem does it solve?
-----------------------------------------------------

Add a handler to interrupt `restic backup` with saving an "interrupted" snapshot

Was the change discussed in an issue or in the forum before?
------------------------------------------------------------

closes #2960 

Checklist
---------

- [x] I have read the [Contribution Guidelines](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#providing-patches)
- [x] I have enabled [maintainer edits for this PR](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork)
- [ ] I have added tests for all changes in this PR
- [ ] I have added documentation for the changes (in the manual)
- [ ] There's a new file in `changelog/unreleased/` that describes the changes for our users (template [here](https://github.com/restic/restic/blob/master/changelog/TEMPLATE))
- [x] I have run `gofmt` on the code in all commits
- [x] All commit messages are formatted in the same style as [the other commits in the repo](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#git-commits)
- [ ] I'm done, this Pull Request is ready for review

### CODE REVIEW

**Comment by rawtaz** [comment]
> While I already have a draft implementation which does exactly the steps I pointed out, I would like to use this issue to discuss the solution.

Why not let this be discussed before opening another PR?

**Comment by aawsome** [comment]
> Why not let this be discussed before opening another PR?

Hey, come on, it's only a draft...

I had to make sure that my idea works in principle before I wanted to propose it. And, once I had it coded, what is the sense in hiding it? Open source is sharing source code, IMHO.

**Comment by rawtaz** [comment]
I know you have good intentions, but when you wrote such a nice and clear issue describing the proposed feature, I'd start with that first. Otherwise there's two numbers (issues and PRs) going up instead of one (issues), taking ever so slightly more of maintainer focus. That's the reason I asked.

File: cmd/restic/cmd_backup.go [hunk 1]
```diff
@@ -601,10 +601,33 @@ func runBackup(opts BackupOptions, gopts GlobalOptions, term *termstatus.Termina
...
+		case <-backupFinish:
+		}
+		return nil
+	})
+
 	if !gopts.JSON {
 		p.V("start backup on %v", targets)
 	}
-	_, id, err := arch.Snapshot(gopts.ctx, targets, snapshotOpts)
+	_, id, err := arch.Snapshot(gopts.ctx, fileCtx, targets, snapshotOpts)
```
Comment (greatroar) [inline cmd/restic/cmd_backup.go:630]: I shudder when seeing two contexts passed around all over the codebase. Surely we can get by with a single one, if need be a custom type (Context being an interface)?

**Comment by aawsome** [comment]
Closing this as IMO #3230 is much more promising.

**Pull request closed without merging**